src/embed_tests/Codecs.cs

```diff
@@ -82,5 +82,130 @@ public class Codecs {
                 Assert.AreEqual(expected: tuple, actual: restored);
             }
         }
+
+        [Test]
+        public void ListCodecTest()
+        {
+            var codec = ListCodec.Instance;
+            var items = new List<PyObject>() { new PyInt(1), new PyInt(2), new PyInt(3) };
+
+            var x = new PyList(items.ToArray());
+            Assert.IsTrue(codec.CanDecode(x, typeof(IList<bool>)));
+            Assert.IsTrue(codec.CanDecode(x, typeof(System.Collections.IEnumerable)));
+            Assert.IsTrue(codec.CanDecode(x, typeof(IEnumerable<int>)));
+            Assert.IsTrue(codec.CanDecode(x, typeof(ICollection<float>)));
+            Assert.IsFalse(codec.CanDecode(x, typeof(bool)));
+
+            //we'd have to copy into a list to do this.  not the best idea to support it.
+            //maybe there can be a flag on listcodec to allow it.
+            Assert.IsFalse(codec.CanDecode(x, typeof(List<int>)));
```

Comment:
@lostmsu @filmor With your idea of multiple codecs I can have a DeepCopyListCodec that takes any ienumerable and converts it by copying to a list.  This is what we do right now in master and if we get rid of that implicit conversion the user can just register this codec if that causes an undesirable behavior change for them.

Reply:
I am thinking, that `Python.Runtime.dll` should only contain codecs, that perform lossless conversions, perhaps even by default. E.g. if a type is exactly `builtins.list` it is safe to convert it to `IList<T>` via some safe wrapper type `PythonList<T>`, that would contain a reference to the original object. Preferably, when this wrapper is passed back to Python it should either just pass the wrapped object, or behave exactly like a normal `list` would.

The codecs, that are not lossless, like converting a `builtins.list` to `T[]`, and codecs, that act on relatively uncommon types like `System.Uri` should go to a separate extra package (created yesterday after seeing your PR): https://github.com/pythonnet/codecs

Reply:
@lostmsu done.  I didn't go as far as to enable by default but that can be done as a followup commit.  Please look closely at the test cases as I tried to capture the behavior of each of the three codecs to make sure it meets your expectation.